src/box/lua/net_replicaset.lua

```diff
@@ -102,6 +102,89 @@ function replicaset_methods:call_leader(func, args, opts)
     end
 end
 
+local watcher_methods = {}
+
+local watcher_mt = {
+    __index = watcher_methods,
+}
+
+-- Check that the first argument is a valid watcher.
+local function watcher_check(watcher, method)
+    if getmetatable(watcher) ~= watcher_mt then
+        local fmt = 'Use watcher:%s(...) instead of watcher.%s(...)'
+        error(string.format(fmt, method, method))
+    end
+end
+
+-- Stop watching the key.
+function watcher_methods.unregister(watcher)
+    watcher_check(watcher, 'unregister')
+    if watcher.handle then
+        watcher.handle:unregister()
+        watcher.handle = nil
+    end
+    watcher.replicaset.leader_watchers[watcher] = nil
+end
+
+-- Get closure with `key`, `value` arguments that calls user watcher function
+-- with extended signature (with the third `instance_name` argument).
+-- In addition the closure uses guard preventing concurrent execution of func.
+local function wrap_watcher_function(watcher, instance)
+    local instance_name = instance.instance_name
+    return function(key, value)
```

Comment:
Dangerous ground here. If we will ever reference `instance` in the callback, we would get another cyclic ref  and break GC. Lets add a test that adding a watch doesn't prevent GC when all public refs are gone.

Reply:
Checked, considering that while the watcher is not unregistered the replicaset must live.